File: backend/app/schemas/requests.py
"""Pydantic v2 schemas for /requests router (Phase 4.1).

Locked spec from `docs/31-api.md` §2 and `docs/02-statuses.md` §7.1.
"""
from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# Position sub-schemas
# ---------------------------------------------------------------------------

class RequestPositionIn(BaseModel):
    """Body for a single position in POST /requests and POST /requests/{id}/positions."""
    name: str = Field(min_length=1)
    qty: float
    unit: Optional[str] = None
    gost_tu: Optional[str] = None
    doc_code: Optional[str] = None


class RequestPositionPatch(BaseModel):
    """Partial update for PATCH /requests/{id}/positions/{pos_id}."""
    name: Optional[str] = None
    qty: Optional[float] = None
    unit: Optional[str] = None
    gost_tu: Optional[str] = None
    doc_code: Optional[str] = None


class PositionOut(BaseModel):
    """Response shape for a single position row."""
    model_config = ConfigDict(from_attributes=True)

    id: int
    parent_id: int
    name: str
    qty: float
    unit: Optional[str] = None
    gost_tu: Optional[str] = None
    doc_code: Optional[str] = None


# ---------------------------------------------------------------------------
# ProcedureOut — minimal shape for /requests/{id} (tender's procedures)
# ---------------------------------------------------------------------------

class ProcedureOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    tender_id: int
    proc: Optional[str] = None
    supplier: Optional[str] = None
    block: str
    status_zakup: Optional[str] = None
    status_postavki: Optional[str] = None
    status_sdelki: Optional[str] = None


# ---------------------------------------------------------------------------
# TenderOut — minimal shape with embedded procedures
# ---------------------------------------------------------------------------

class TenderOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    parent_id: int
    num: Optional[str] = None
    procedures: list[ProcedureOut] = []


# ---------------------------------------------------------------------------
# Request create / patch
# ---------------------------------------------------------------------------

class RequestCreate(BaseModel):
    """Body for POST /requests — create a parent + positions."""
    code: str = Field(min_length=1)
    title: str = Field(min_length=1)
    mtr: Optional[str] = None
    srok: Optional[str] = None
    dept: Optional[str] = None
    positions: list[RequestPositionIn] = []


class RequestPatch(BaseModel):
    """Body for PATCH /requests/{id} — partial update."""
    title: Optional[str] = None
    mtr: Optional[str] = None
    srok: Optional[str] = None
    dept: Optional[str] = None


class RequestDuplicate(BaseModel):
    """Body for POST /requests/{id}/duplicate."""
    code: str = Field(min_length=1)


# ---------------------------------------------------------------------------
# Response shapes
# ---------------------------------------------------------------------------

class RequestOut(BaseModel):
    """Full request (with embedded positions + tenders) — used for GET/{id},
    POST, PATCH, cancel/uncancel/duplicate responses."""
    model_config = ConfigDict(from_attributes=True)

    id: int
    code: str
    title: str
    mtr: Optional[str] = None
    srok: Optional[str] = None
    zagruzka: str
    sostavitel: str
    created_by: Optional[int] = None
    dept: Optional[str] = None
    status: str
    created_at: str
    positions: list[PositionOut] = []
    tenders: list[TenderOut] = []


class RequestListItem(BaseModel):
    """Shorter shape for the list endpoint (GET /requests)."""
    model_config = ConfigDict(from_attributes=True)

    id: int
    code: str
    title: str
    mtr: Optional[str] = None
    srok: Optional[str] = None
    zagruzka: str
    sostavitel: str
    status: str
    created_at: str
    position_count: int


class PaginatedRequests(BaseModel):
    items: list[RequestListItem]
    total: int


__all__ = [
    "RequestPositionIn",
    "RequestPositionPatch",
    "PositionOut",
    "ProcedureOut",
    "TenderOut",
    "RequestCreate",
    "RequestPatch",
    "RequestDuplicate",
    "RequestOut",
    "RequestListItem",
    "PaginatedRequests",
]

File: frontend/src/api/requests.ts
import { apiFetch } from './client'

// ---- Types ----------------------------------------------------------------

export type RequestStatus = 'awaiting' | 'cancelled'
export type RequestSort = 'created_at' | 'code' | 'title'

export type RequestListItem = {
  id: number
  code: string
  title: string
  mtr: string | null
  srok: string | null
  zagruzka: string
  sostavitel: string
  status: RequestStatus
  created_at: string
  position_count: number
}

export type PaginatedRequests = {
  items: RequestListItem[]
  total: number
}

export type RequestPosition = {
  id: number
  parent_id: number
  name: string
  qty: number
  unit: string | null
  gost_tu: string | null
  doc_code: string | null
}

export type ProcedureOut = {
  id: number
  tender_id: number
  proc: string | null
  supplier: string | null
  block: string
  status_zakup: string | null
  status_postavki: string | null
  status_sdelki: string | null
}

export type TenderOut = {
  id: number
  parent_id: number
  num: string | null
  procedures: ProcedureOut[]
}

export type RequestOut = {
  id: number
  code: string
  title: string
  mtr: string | null
  srok: string | null
  zagruzka: string
  sostavitel: string
  created_by: number | null
  dept: string | null
  status: RequestStatus
  created_at: string
  positions: RequestPosition[]
  tenders: TenderOut[]
}

export type RequestPositionInput = {
  name: string
  qty: number
  unit?: string | null
  gost_tu?: string | null
  doc_code?: string | null
}

export type RequestCreate = {
  code: string
  title: string
  mtr?: string | null
  srok?: string | null
  dept?: string | null
  positions: RequestPositionInput[]
}

export type RequestPatch = {
  title?: string
  mtr?: string | null
  srok?: string | null
  dept?: string | null
}

export type TakeToWorkResponse = {
  tender_id: number
  procedure_id: number
}

// ---- Helpers --------------------------------------------------------------

function buildQuery(params: Record<string, unknown>): string {
  const usp = new URLSearchParams()
  for (const [k, v] of Object.entries(params)) {
    if (v === undefined || v === null || v === '') continue
    usp.set(k, String(v))
  }
  const s = usp.toString()
  return s ? `?${s}` : ''
}

// ---- Endpoints ------------------------------------------------------------

export function listRequests(params: {
  include_archived?: boolean
  status?: RequestStatus
  search?: string
  sort?: RequestSort
  page?: number
  page_size?: number
} = {}): Promise<PaginatedRequests> {
  return apiFetch<PaginatedRequests>(`/requests${buildQuery(params)}`)
}

export function getRequest(id: number): Promise<RequestOut> {
  return apiFetch<RequestOut>(`/requests/${id}`)
}

export function createRequest(payload: RequestCreate): Promise<RequestOut> {
  return apiFetch<RequestOut>('/requests', { method: 'POST', body: payload })
}

export function patchRequest(
  id: number,
  payload: RequestPatch,
): Promise<RequestOut> {
  return apiFetch<RequestOut>(`/requests/${id}`, {
    method: 'PATCH',
    body: payload,
  })
}

export function cancelRequest(id: number): Promise<RequestOut> {
  return apiFetch<RequestOut>(`/requests/${id}/cancel`, { method: 'POST' })
}

export function uncancelRequest(id: number): Promise<RequestOut> {
  return apiFetch<RequestOut>(`/requests/${id}/uncancel`, { method: 'POST' })
}

export function duplicateRequest(id: number, code: string): Promise<RequestOut> {
  return apiFetch<RequestOut>(`/requests/${id}/duplicate`, {
    method: 'POST',
    body: { code },
  })
}

export function takeToWork(id: number): Promise<TakeToWorkResponse> {
  return apiFetch<TakeToWorkResponse>(`/requests/${id}/take-to-work`, {
    method: 'POST',
  })
}

export function listPositions(requestId: number): Promise<RequestPosition[]> {
  return apiFetch<RequestPosition[]>(`/requests/${requestId}/positions`)
}

export function addPositions(
  requestId: number,
  positions: RequestPositionInput[],
): Promise<RequestPosition[]> {
  return apiFetch<RequestPosition[]>(`/requests/${requestId}/positions`, {
    method: 'POST',
    body: positions,
  })
}

export function patchPosition(
  requestId: number,
  posId: number,
  payload: Partial<RequestPosition>,
): Promise<RequestPosition> {
  return apiFetch<RequestPosition>(
    `/requests/${requestId}/positions/${posId}`,
    { method: 'PATCH', body: payload },
  )
}

export function deletePosition(
  requestId: number,
  posId: number,
): Promise<void> {
  return apiFetch<void>(`/requests/${requestId}/positions/${posId}`, {
    method: 'DELETE',
  })
}
